tt_metal/api/tt-metalium/maybe_remote.hpp

```diff
@@ -0,0 +1,242 @@
+
+    // Helper for throwing remote access errors
+    [[noreturn]] static void throw_remote_access_error();
+
+public:
+    // No default constructor - must be explicit about remote/local
+    MaybeRemote() = delete;
+
+    // Named constructors for clarity
+    static MaybeRemote local(T value);
+    static MaybeRemote remote();
+
+    // Query methods
+    [[nodiscard]] bool is_local() const noexcept;
+    [[nodiscard]] bool is_remote() const noexcept;
+
+    // Access methods: throw if remote, return local value if local.
+    [[nodiscard]] T& value() &;
+    [[nodiscard]] const T& value() const&;
+
+    // Pointer-like operators for convenience
+    [[nodiscard]] T& operator*() &;
+    [[nodiscard]] const T& operator*() const&;
+    [[nodiscard]] T* operator->();
+    [[nodiscard]] const T* operator->() const;
+
+    // Simple pattern matching for local/remote cases
+    template <typename LocalFunc, typename RemoteFunc>
+    [[nodiscard]] auto when(LocalFunc&& on_local, RemoteFunc&& on_remote) const -> decltype(auto);
+
+    // Equality operators
+    bool operator==(const MaybeRemote& other) const = default;
+
+    // Debugging support
+    [[nodiscard]] std::string to_string() const;
+
+};
+
+// Type aliases for common use cases
+using MaybeRemoteDeviceId = MaybeRemote<int>;
+using MaybeRemoteDevice = MaybeRemote<IDevice*>;
+
+// ============================================================================
+// MaybeRemote Implementation
+// ============================================================================
+
+template <typename T>
+MaybeRemote<T>::MaybeRemote(RemoteDevice) : value_(RemoteDevice{}) {}
+
+template <typename T>
+MaybeRemote<T>::MaybeRemote(T value) : value_(std::move(value)) {}
+
+template <typename T>
+[[noreturn]] void MaybeRemote<T>::throw_remote_access_error() {
+    TT_THROW("Attempted to access remote device from process that does not have it!");
+}
+
+template <typename T>
+MaybeRemote<T> MaybeRemote<T>::local(T value) {
+    return MaybeRemote(std::move(value));
+}
+
+template <typename T>
+MaybeRemote<T> MaybeRemote<T>::remote() {
+    return MaybeRemote(RemoteDevice{});
+}
+
+template <typename T>
+bool MaybeRemote<T>::is_local() const noexcept {
+    return std::holds_alternative<T>(value_);
+}
+
+template <typename T>
+bool MaybeRemote<T>::is_remote() const noexcept {
+    return std::holds_alternative<RemoteDevice>(value_);
+}
+
+template <typename T>
+T& MaybeRemote<T>::value() & {
+    if (is_remote()) {
+        throw_remote_access_error();
+    }
+    return std::get<T>(value_);
+}
+
+template <typename T>
+const T& MaybeRemote<T>::value() const& {
+    if (is_remote()) {
+        throw_remote_access_error();
+    }
+    return std::get<T>(value_);
+}
+
+template <typename T>
+T& MaybeRemote<T>::operator*() & {
+    return value();
+}
+
+template <typename T>
+const T& MaybeRemote<T>::operator*() const& {
+    return value();
+}
+
+template <typename T>
+T* MaybeRemote<T>::operator->() {
```

Comment:
```
template <typename T>
T* MaybeRemote<T>::operator->() {
    if (auto* p = std::get_if<T>(&value_); p) {
        return p;                     // local -– hand back the pointer
    }
    throw_remote_access_error();      // remote – raise
}
```

Looks a little bit better

Also please don't forget to add nodiscard. cc @omilyutin-tt 
I dont want to enforce it everwhere in the ttnn code but for generic types I thnk it is a must.